src/traces/heatmap/plot.js

```diff
@@ -18,32 +19,28 @@ var xmlnsNamespaces = require('../../constants/xmlns_namespaces');
 
 var maxRowLength = require('./max_row_length');
 
+module.exports = function(gd, plotinfo, cdheatmaps, heatmapLayer) {
+    var i;
 
-module.exports = function(gd, plotinfo, cdheatmaps) {
-    for(var i = 0; i < cdheatmaps.length; i++) {
-        plotOne(gd, plotinfo, cdheatmaps[i]);
+    heatmapLayer.selectAll('.hm > image').each(function(d) {
+        var oldTrace = d.trace || {};
```

Comment:
when is there no `d.trace`?

Reply:
when a heatmap trace is completely off screen:

https://github.com/plotly/plotly.js/blob/bab2a5b692a21200e295c3909056818591e839ff/src/traces/heatmap/plot.js#L138-L142

Reply:
This is not a big deal, but wouldn't an off-screen heatmap just not contribute to the `'.hm > image'` selection at all?

Reply:
removing that `|| {}` makes

https://github.com/plotly/plotly.js/blob/c0b38dab5044310468355bfa533ccad516399f87/test/jasmine/tests/heatmap_test.js#L482-L506

fail with

![image](https://user-images.githubusercontent.com/6675409/39434761-7e5c9e02-4c67-11e8-94c4-48eb7e54e919.png)

Reply:
... I didn't investigate further.

Reply:
Alright, not what my quick glance at the code suggested but you're right, no need to dig in any further.